File: algorithms/2_medium/F/00890_find_and_replace_pattern/solution.cpp
class Solution {
    static inline string custom_transform(const string& word) {
        char curr_available{ 'a' };
        array<char, 26> char_map{};
        string res;
        res.reserve(word.size());

        for (const auto& ch : word) {
            const size_t index{ static_cast<size_t>(ch - 'a') };

            if (char_map[index] == 0) {
                char_map[index] = curr_available;
                curr_available += 1;
            }

            res.push_back(char_map[index]);
        }

        return res;
    }
public:
    static vector<string> findAndReplacePattern(const vector<string>& words, const string& pattern) {
        const auto pattern_transformed{ custom_transform(pattern) };
        vector<string> res{};

        for (const auto& word : words) {
            const auto word_transformed{ custom_transform(word) };

            if (word_transformed == pattern_transformed) {
                res.push_back(word);
            }
        }

        return res;
    }
};

File: algorithms/2_medium/F/00890_find_and_replace_pattern/solution.rs
impl Solution {

    #[inline(always)]
    fn custom_transform(bytes: &[u8]) -> Vec<u8> {
        let mut curr_available = b'a';
        let mut byte_map = [0_u8; 26];
        let mut res: Vec<u8> = Vec::with_capacity(bytes.len());

        for byte in bytes.iter() {
            let byte_map_index = (*byte - b'a') as usize;

            if byte_map[byte_map_index] == 0 {
                byte_map[byte_map_index] = curr_available;
                curr_available += 1;
            }

            res.push(byte_map[byte_map_index]);
        }

        res
    }

    pub fn find_and_replace_pattern(words: Vec<String>, pattern: String) -> Vec<String> {
        let pattern_bytes_transformed = Self::custom_transform(pattern.as_bytes());
        let mut res: Vec<String> = Vec::new();

        for word in words.into_iter() {
            let word_bytes_transformed = Self::custom_transform(word.as_bytes());

            if word_bytes_transformed == pattern_bytes_transformed {
                res.push(word);
            }
        }

        res
    }
}
